### ISSUE TITLE
NPE while compiling an LMF-capable SAM inside an anonymous JS class

### ISSUE THREAD

**Issue by machaval** [scala-js/scala-js#3264]
```
Caused by: java.lang.NullPointerException
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.synthesizeSAMWrapper(GenJSCode.scala:5111)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genAnonFunction(GenJSCode.scala:5100)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1948)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2977)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1757)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genBlock$8(GenJSCode.scala:2976)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2976)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genLabelApply(GenJSCode.scala:2436)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2244)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3033)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genTranslatedMatch$3(GenJSCode.scala:3037)
        at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:739)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:738)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genTranslatedMatch(GenJSCode.scala:3017)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.translateMatch$1(GenJSCode.scala:2945)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2952)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2977)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genLabelApply(GenJSCode.scala:2436)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2244)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3033)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3027)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genCaseBody$1(GenJSCode.scala:3023)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genTranslatedMatch$3(GenJSCode.scala:3037)
        at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:739)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:738)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genTranslatedMatch(GenJSCode.scala:3017)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.translateMatch$1(GenJSCode.scala:2945)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2952)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBody$1(GenJSCode.scala:1683)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodDef(GenJSCode.scala:1688)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$8(GenJSCode.scala:1499)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$1(GenJSCode.scala:1461)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodWithCurrentLocalNameScope(GenJSCode.scala:1406)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethod$1(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope(JSEncoding.scala:54)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope$(JSEncoding.scala:50)
        at org.scalajs.core.compiler.GenJSCode.withNewLocalNameScope(GenJSCode.scala:31)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethod(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:380)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3$adapted(GenJSCode.scala:371)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genClass(GenJSCode.scala:386)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genApplyNew$6(GenJSCode.scala:2398)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope(JSEncoding.scala:54)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope$(JSEncoding.scala:50)
        at org.scalajs.core.compiler.GenJSCode.withNewLocalNameScope(GenJSCode.scala:31)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$nestedGenerateClass$1(GenJSCode.scala:181)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.nestedGenerateClass(GenJSCode.scala:181)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genApplyNew$5(GenJSCode.scala:2398)
        at scala.Option.getOrElse(Option.scala:121)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApplyNew(GenJSCode.scala:2395)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2240)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2977)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1894)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genActualArgs$5(GenJSCode.scala:4526)
        at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:739)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:738)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genActualArgs(GenJSCode.scala:4517)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genNormalApply(GenJSCode.scala:2545)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2256)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genNormalApply(GenJSCode.scala:2545)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2256)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genIsAsInstanceOf(GenJSCode.scala:2288)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApplyTypeApply(GenJSCode.scala:2276)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genApply(GenJSCode.scala:2231)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1800)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1757)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStat(GenJSCode.scala:1709)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genBlock$8(GenJSCode.scala:2976)
        at scala.collection.immutable.List.map(List.scala:283)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBlock(GenJSCode.scala:2976)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genStatOrExpr(GenJSCode.scala:1888)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genExpr(GenJSCode.scala:1728)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genBody$1(GenJSCode.scala:1683)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodDef(GenJSCode.scala:1688)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$8(GenJSCode.scala:1499)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethodWithCurrentLocalNameScope$1(GenJSCode.scala:1461)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethodWithCurrentLocalNameScope(GenJSCode.scala:1406)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genMethod$1(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope(JSEncoding.scala:54)
        at org.scalajs.core.compiler.JSEncoding.withNewLocalNameScope$(JSEncoding.scala:50)
        at org.scalajs.core.compiler.GenJSCode.withNewLocalNameScope(GenJSCode.scala:31)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genMethod(GenJSCode.scala:1371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:380)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$genClass$3$adapted(GenJSCode.scala:371)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.gen$1(GenJSCode.scala:371)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.genClass(GenJSCode.scala:386)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$apply$5(GenJSCode.scala:302)
        at org.scalajs.core.compiler.util.ScopedVar$.withScopedVars(ScopedVar.scala:35)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.$anonfun$apply$4(GenJSCode.scala:289)
        at scala.collection.immutable.List.foreach(List.scala:389)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.apply(GenJSCode.scala:276)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:436)
        at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:429)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:400)
        at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:400)
        at scala.collection.Iterator.foreach(Iterator.scala:929)
        at scala.collection.Iterator.foreach$(Iterator.scala:929)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
        at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:400)
        at org.scalajs.core.compiler.GenJSCode$JSCodePhase.run(GenJSCode.scala:217)
        at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1452)
        at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1436)
        at scala.tools.nsc.Global$Run.compileSources(Global.scala:1429)
        at scala.tools.nsc.Global$Run.compile(Global.scala:1545)
        at xsbt.CachedCompiler0.run(CompilerInterface.scala:116)
        at xsbt.CachedCompiler0.run(CompilerInterface.scala:95)
        at xsbt.CompilerInterface.run(CompilerInterface.scala:26)
        at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:101)
        at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:47)
        at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:41)
        at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply$mcV$sp(AggressiveCompile.scala:97)
        at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply(AggressiveCompile.scala:97)
        at sbt.compiler.AggressiveCompile$$anonfun$3$$anonfun$compileScala$1$1.apply(AggressiveCompile.scala:97)
        at sbt.compiler.AggressiveCompile.sbt$compiler$AggressiveCompile$$timed(AggressiveCompile.scala:162)
        at sbt.compiler.AggressiveCompile$$anonfun$3.compileScala$1(AggressiveCompile.scala:96)
        at sbt.compiler.AggressiveCompile$$anonfun$3.apply(AggressiveCompile.scala:139)
        at sbt.compiler.AggressiveCompile$$anonfun$3.apply(AggressiveCompile.scala:86)
        at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:38)
        at sbt.inc.IncrementalCompile$$anonfun$doCompile$1.apply(Compile.scala:36)
        at sbt.inc.IncrementalCommon.cycle(IncrementalCommon.scala:31)
        at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:39)
        at sbt.inc.Incremental$$anonfun$1.apply(Incremental.scala:38)
        at sbt.inc.Incremental$.manageClassfiles(Incremental.scala:66)
        at sbt.inc.Incremental$.compile(Incremental.scala:38)
        at sbt.inc.IncrementalCompile$.apply(Compile.scala:26)
        at sbt.compiler.AggressiveCompile.compile2(AggressiveCompile.scala:153)
        at sbt.compiler.AggressiveCompile.compile1(AggressiveCompile.scala:70)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:201)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:183)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:174)
        at com.typesafe.zinc.Compiler.compile(Compiler.scala:165)
        at org.gradle.api.internal.tasks.scala.ZincScalaCompiler$Compiler.execute(ZincScalaCompiler.java:92)
        at org.gradle.api.internal.tasks.scala.ZincScalaCompiler.execute(ZincScalaCompiler.java:66)
        at org.gradle.api.internal.tasks.scala.ZincScalaCompiler.execute(ZincScalaCompiler.java:48)
        at org.gradle.api.internal.tasks.compile.daemon.CompilerDaemonServer.execute(CompilerDaemonServer.java:31)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:87)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:377)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)

```

**Comment by machaval** [comment]
Found the root cause 

```
            rt.copy(referenceResolver = () => {
              recursionDetector.resolve(rt, (referencedType) => treeMap(referencedType, callback, recursionDetector))
            })
```
And is fixed if I changed the outer lambda with the
```
  val resolver = new ReferenceTypeResolver() {
              override def resolveType() = {
                recursionDetector.resolve(rt, (referencedType) => treeMap(referencedType, callback, recursionDetector))
              }
            }
            rt.copy(referenceResolver = resolver)
          }
```

**Comment by sjrd** [comment]
Could you please provide a code snippet that reproduces the issue when you try to compile it?

Also, when you mention code from the repo, could you please link to the relevant line in the relevant file?

**Comment by sjrd** [comment]
The following code reproduces the bug:
```scala
abstract class Parent extends js.Object {
  def foo(x: Int): Int
}

trait MySAM {
  def bar(x: Int): Int
}

object HelloWorld {
  def main(args: Array[String]): Unit = {
    val outer = new Parent {
      def foo(x: Int): Int = {
        val innerSAM: MySAM = _ * 2
        innerSAM.bar(x + 3)
      }
    }

    assert(16 == outer.foo(5))
  }
}
```
Minimized conditions: an LMF-capable SAM inside an anonymous JS class.

**Comment by machaval** [comment]
Thanks for the quick fix!!!! Sorry I was out without my laptop and was not able to provide the full example.

**Comment by sjrd** [comment]
Fixed in b5e6e62fd19ec6771b3003ad1f7eed9c3bce70ea.
